common/src/node.rs

```diff
@@ -87,34 +87,43 @@ impl Graph {
         Some((chunk, coords))
     }
 
-    /// Populates a chunk with the given voxel data and ensures that margins are correctly cleared if necessary.
-    pub fn populate_chunk(&mut self, chunk: ChunkId, mut new_data: VoxelData, modified: bool) {
-        // New solid chunks should have their margin cleared if they are adjacent to any modified chunks.
-        // See the function description of VoxelData::clear_margin for why this is necessary.
-        if new_data.is_solid() {
-            // Loop through all six potential chunk neighbors. If any are modified, the `new_data` should have
-            // its margin cleared.
-            for chunk_direction in ChunkDirection::iter() {
-                if let Some(chunk_id) =
-                    self.get_chunk_neighbor(chunk, chunk_direction.axis, chunk_direction.sign)
+    /// Populates a chunk with the given voxel data and ensures that margins are correctly fixed up if necessary.
+    pub fn populate_chunk(&mut self, chunk: ChunkId, mut voxels: VoxelData, modified: bool) {
+        let dimension = self.layout().dimension;
+        // Fix up margins for the chunk we're inserting along with any neighboring chunks
+        for chunk_direction in ChunkDirection::iter() {
+            if let Some(Chunk::Populated {
+                modified: neighbor_modified,
+                voxels: neighbor_voxels,
+                surface: neighbor_surface,
+                old_surface: neighbor_old_surface,
+            }) = self
+                .get_chunk_neighbor(chunk, chunk_direction.axis, chunk_direction.sign)
+                .map(|chunk_id| &mut self[chunk_id])
+            {
+                // We need to fix up margins between the current chunk and the neighboring chunk if and only if
+                // there's a potential surface between them. This can occur if either is modified or if neither
+                // is designated as solid. Note that if one is designated as solid, that means that it's deep enough
+                // in the terrain or up in the air that there will be no surface between them.
```

Comment:
> if one is designated as solid, that means that it's deep enough in the terrain or up in the air that there will be no surface between them.

Is this necessarily true? What if we have a chunk whose interior is solid void, adjacent to a chunk (solid or not) that contains solid voxels incident to the common face?

Reply:
That's a good question. My understanding is that `VoxelData::Solid` corresponds to chunks that are solid, including their margins. I vaguely remember that being explicitly stated somewhere in the past, but I don't remember where.

However, I also vaguely remember that this extra restriction only applied to non-void solid chunks because the purpose of that restriction was to be able to say that solid chunks don't need to be rendered at all.

I would propose that we make this apply to void chunks as well, since that would allow us to make this assumption.

Reply:
> However, I also vaguely remember that this extra restriction only applied to non-void solid chunks because the purpose of that restriction was to be able to say that solid chunks don't need to be rendered at all.

Skimming the worldgen logic supports this hypothesis: we generate solid dirt chunks or solid void chunks when distant from the surface, and otherwise start with a solid void chunk and possibly fill in some voxels. Hard to say for sure how exact "distant from the surface" is, though.

> I would propose that we make this apply to void chunks as well, since that would allow us to make this assumption.

Meaning we convert solid void chunks to dense chunks if their margins are touched? Sure, that seems reasonable.

Reply:
In that case, I believe this PR is in a good state. Unless the logic was done with a razor-thin margin, worldgen already ensures that solid void chunks don't have any voxels directly adjacent to them since the call to `const ELEVATION_MARGIN: f64 = 0.7;` was empirically determined.

However, I'll see if I can find the discussion I vaguely recall about the meaning of `VoxelData::Solid`. I'll leave this comment thread unresolved until you confirm that you're happy with the current state of things.